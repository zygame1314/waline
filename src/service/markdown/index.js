const { katex: katexPlugin } = require('@mdit/plugin-katex');
const { sub: subPlugin } = require('@mdit/plugin-sub');
const { sup: supPlugin } = require('@mdit/plugin-sup');
const MarkdownIt = require('markdown-it');
const emojiPlugin = require('markdown-it-emoji');

const { resolveHighlighter } = require('./highlight.js');
const { mathjaxPlugin } = require('./mathjax.js');
const { sanitize } = require('./xss.js');

const getMarkdownParser = () => {
  const { markdown = {} } = think.config();
  const { config = {}, plugin = {} } = markdown;

  // markdown-it instance
  const markdownIt = MarkdownIt({
    breaks: true,
    linkify: true, // Auto convert URL-like text to links
    typographer: true, // Enable some language-neutral replacement + quotes beautification

    // default highlight
    highlight: (code, lang) => {
      const highlighter = resolveHighlighter(lang);

      return highlighter ? highlighter(code) : '';
    },

    ...config,

    // should always enable html option due to parsed emoji
    html: true,
  });

  const { emoji, tex, mathjax, katex, sub, sup } = plugin;

  // parse emoji
  if (emoji !== false) {
    markdownIt.use(emojiPlugin.full, typeof emoji === 'object' ? emoji : {});
  }

  // parse sub
  if (sub !== false) {
    markdownIt.use(subPlugin);
  }

  // parse sup
  if (sup !== false) {
    markdownIt.use(supPlugin);
  }

  // parse tex
  if (tex === 'katex') {
    markdownIt.use(katexPlugin, {
      ...katex,
      output: 'mathml',
    });
  } else if (tex !== false) {
    markdownIt.use(mathjaxPlugin, mathjax);
  }

  return (content) => sanitize(markdownIt.render(content));
};

module.exports = { getMarkdownParser };
