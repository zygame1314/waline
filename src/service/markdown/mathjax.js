const { liteAdaptor } = require('mathjax-full/js/adaptors/liteAdaptor.js');
const { RegisterHTMLHandler } = require('mathjax-full/js/handlers/html.js');
const { AllPackages } = require('mathjax-full/js/input/tex/AllPackages.js');
const { TeX } = require('mathjax-full/js/input/tex.js');
const { mathjax } = require('mathjax-full/js/mathjax');
const { SVG } = require('mathjax-full/js/output/svg.js');

const { inlineTeX, blockTeX } = require('./mathCommon');
const { escapeHtml } = require('./utils');

const mathjaxPlugin = (md) => {
  const adaptor = liteAdaptor();

  RegisterHTMLHandler(adaptor);

  const packages = AllPackages.sort();
  const tex = new TeX({ packages });
  const svg = new SVG({ fontCache: 'none' });

  const texToNode = mathjax.document('', { InputJax: tex, OutputJax: svg });

  const inline = (tex) => {
    const node = texToNode.convert(tex, { display: false });
    let svg = adaptor.innerHTML(node);

    if (svg.includes('data-mml-node="merror"')) {
      const errorTitle = svg.match(/<title>(.*?)<\/title>/)[1];

      svg = `<span class='mathjax-error' title='${escapeHtml(
        errorTitle,
      )}'>${escapeHtml(tex)}</span>`;
    }

    return svg;
  };

  const block = (tex) => {
    const node = texToNode.convert(tex, { display: true });
    let svg = adaptor.innerHTML(node);

    if (svg.includes('data-mml-node="merror"')) {
      const errorTitle = svg.match(/<title>(.*?)<\/title>/)[1];

      svg = `<p class='mathjax-block mathjax-error' title='${escapeHtml(
        errorTitle,
      )}'>${escapeHtml(tex)}</p>`;
    } else {
      svg = svg.replace(/(width=".*?")/, 'width="100%"');
    }

    return svg;
  };

  md.inline.ruler.after('escape', 'inlineTeX', inlineTeX);

  // Itâ€™s a workaround here because types issue
  md.block.ruler.after('blockquote', 'blockTeX', blockTeX, {
    alt: ['paragraph', 'reference', 'blockquote', 'list'],
  });

  md.renderer.rules.inlineTeX = (tokens, idx) => inline(tokens[idx].content);

  md.renderer.rules.blockTeX = (tokens, idx) =>
    `${block(tokens[idx].content)}\n`;
};

module.exports = {
  mathjaxPlugin,
};
