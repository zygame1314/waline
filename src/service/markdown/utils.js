const escapeHtml = (unsafeHTML) =>
  unsafeHTML
    .replace(/&/gu, '&amp;')
    .replace(/</gu, '&lt;')
    .replace(/>/gu, '&gt;')
    .replace(/"/gu, '&quot;')
    .replace(/'/gu, '&#039;');

module.exports = {
  escapeHtml,
};
