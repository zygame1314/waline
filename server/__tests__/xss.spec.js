import MarkdownIt from 'markdown-it';
import { describe, expect, it } from 'vitest';

import { sanitize } from '../src/service/markdown/xss';

const parser = (content) =>
  sanitize(new MarkdownIt({ html: true }).render(content));

describe('XSS test', () => {
  it('Should render', () => {
    const content = `
Waline is a good framework. :money:

## 特性

- 快速
- 真·安全
- Markdown 语法支持
- 轻量易用
- 免费部署
- 多种部署部署方式和存储服务支持，每列选择一项多达 48 种部署方式任君选择

|                | Waline         |              |
| -------------- | -------------- | ------------ |
| **客户端脚本** | **服务端部署** | **数据存储** |
| @waline/client | Vercel         | LeanCloud    |
| MiniValine     | CloudBase      | CloudBase    |
|                | Docker         | MongoDB      |
|                | 独立部署       | MySQL        |
|                |                | SQLite       |
|                |                | PostgreSQL   |
|                |                | GitHub       |
`;

    expect(parser(content)).toMatchSnapshot();
  });

  it('Should protect', () => {
    expect(parser(`<img src="x" onerror="alert('img')">`)).toEqual(
      '<img src="x">',
    );
    expect(parser('<script>alert("hello world")</script>')).toEqual('');

    expect(
      parser(
        '<p>Waline is <iframe//src=jaVa&Tab;script:alert(3)></iframe>awesome</p>',
      ),
    ).toEqual('<p>Waline is awesome</p>');

    expect(
      parser('<p>Waline is <iframe//src=jaVa&Tab;script:alert(3)>awesome</p>'),
    ).toEqual('<p>Waline is </p>');
  });

  it('Should resolve unmatching html tags', () => {
    expect(parser('<TABLE><tr><td>HELLO</a></TAB>\n<p>Waline</p>')).toEqual(
      '<table><tbody><tr><td>HELLO\n<p>Waline</p></td></tr></tbody></table>',
    );
  });

  it('Should not autoplay or preload media', () => {
    expect(parser('<audio autoplay preload="auto" src="x">')).toEqual(
      '<audio src="x" preload="none"></audio>',
    );
    expect(parser('<audio autoplay src="x"></audio>')).toEqual(
      '<p><audio src="x" preload="none"></audio></p>\n',
    );

    expect(parser('<video autoplay preload="auto" src="x">')).toEqual(
      '<video src="x" preload="none"></video>',
    );
    expect(parser('<video autoplay src="x"></video>')).toEqual(
      '<p><video src="x" preload="none"></video></p>\n',
    );
  });

  it('Should resolve links', () => {
    expect(parser('[link](https://example.com)')).toEqual(
      '<p><a href="https://example.com" target="_blank" rel="nofollow noreferrer noopener">link</a></p>\n',
    );
    expect(
      parser(
        '<p><a href="https://example.com" rel="opener prefetch">link</a></p>',
      ),
    ).toEqual(
      '<p><a rel="nofollow noreferrer noopener" href="https://example.com" target="_blank">link</a></p>',
    );
  });

  it('Should forbid forms and inputs', () => {
    expect(parser('<form></form>')).toEqual('');
    expect(parser('<input type="password" autocomplete>')).toEqual('');
  });

  it('Should forbid style', () => {
    expect(
      parser(
        '<div style="position:fixed;top:0;left:0;width:100vh;height:100vh;">广告文字</div>',
      ),
    ).toEqual('<div>广告文字</div>');
    expect(
      parser(
        '<div id="ad">广告文字</div><style>#ad{position:fixed;top:0;left:0;width:100vh;height:100vh;}</style>',
      ),
    ).toEqual('<div id="ad">广告文字</div>');
  });
});
