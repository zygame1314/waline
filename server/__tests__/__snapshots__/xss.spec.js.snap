// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`XSS test > Should render 1`] = `
"<p>Waline is a good framework. :money:</p>
<h2>特性</h2>
<ul>
<li>快速</li>
<li>真·安全</li>
<li>Markdown 语法支持</li>
<li>轻量易用</li>
<li>免费部署</li>
<li>多种部署部署方式和存储服务支持，每列选择一项多达 48 种部署方式任君选择</li>
</ul>
<table>
<thead>
<tr>
<th></th>
<th>Waline</th>
<th></th>
</tr>
</thead>
<tbody>
<tr>
<td><strong>客户端脚本</strong></td>
<td><strong>服务端部署</strong></td>
<td><strong>数据存储</strong></td>
</tr>
<tr>
<td>@waline/client</td>
<td>Vercel</td>
<td>LeanCloud</td>
</tr>
<tr>
<td>MiniValine</td>
<td>CloudBase</td>
<td>CloudBase</td>
</tr>
<tr>
<td></td>
<td>Docker</td>
<td>MongoDB</td>
</tr>
<tr>
<td></td>
<td>独立部署</td>
<td>MySQL</td>
</tr>
<tr>
<td></td>
<td></td>
<td>SQLite</td>
</tr>
<tr>
<td></td>
<td></td>
<td>PostgreSQL</td>
</tr>
<tr>
<td></td>
<td></td>
<td>GitHub</td>
</tr>
</tbody>
</table>
"
`;
